abel/linbasex.py

```diff
@@ -0,0 +1,418 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import numpy as np
+import abel
+import scipy as sci
+from scipy.special import eval_legendre
+from scipy import ndimage
+
+###############################################################################
+# linbasex - inversion procedure based on 1-dimensional projections of
+#            velocity-map images
+#
+# As described in:
+#   Gerber, Thomas, Yuzhu Liu, Gregor Knopp, Patrick Hemberger, Andras Bodi,
+#   Peter Radi, and Yaroslav Sych,
+#     “Charged Particle Velocity Map Image Reconstruction with One-Dimensional
+#      Projections of Spherical Functions.”
+#     Review of Scientific Instruments 84, no. 3 (March 1, 2013):
+#                                      033101–033101 – 10.
+#     doi:10.1063/1.4793404.
+#
+# 2016-04- Thomas Gerber and Daniel Hickstein - theory and code updates
+# 2016-04- Stephen Gibson core code extracted from the supplied jupyter
+#          notebook (see #167: https://github.com/PyAbel/PyAbel/issues/167)
+#
+###############################################################################
+
+_linbasex_parameter_docstring = \
+    r"""Inverse Abel transform using 1d projections of VM-images.
+
+    `Gerber, Thomas, Yuzhu Liu, Gregor Knopp, Patrick Hemberger, Andras Bodi,
+    Peter Radi, and Yaroslav Sych.
+    Charged Particle Velocity Map Image Reconstruction with One-Dimensional Projections of Spherical Functions.` Review of Scientific Instruments 84, no. 3 (March 1, 2013): 033101–033101 – 10.
+    <http://dx.doi.org/10.1063/1.4793404>`_
+
+     linbasex determines the contribution of spherical functions,
+     calculating the weight of each :math:`Y_{i0}`. The reconstructed 3D object
+     is obtained by adding all the :math:`Y_{i0}` contributions, from which
+     slices are derived.
+
+
+    Parameters
+    ----------
+    Dat: numpy 2D array
+        image data must be square shape of odd size
+    an: list
+        projection angle, in degrees
+        e.g. [0, 90] or [0, 54.7356, 90] or [0, 45, 90, 135]
+    un: list
+        order of Legendre polynomials to be used as the expansion
+            even polynomials [0, 2, ...] gerade
+            odd polynomials [1, 3, ...] ungerade
+            all orders [0, 1, 2, ...].
+        In a single photon experiment there are only anisotropies up to
+        second order. The interaction of 4 photons (four wave mixing) yields
+        anisotropies up to order 8.
+    inc: int
+        number of pixels per Newton sphere (default 1)
+    sig_s: float
+        sigma for smoothing (default 0.5)
+    rcond: float
+        (default 0.0005) scipy.linalg.lstsq fit conditioning value.
```

Comment:
In the presence of noise the equation system is often ill posed and needs some conditioning. In some VMI-evaluation software a conditioning for solving the linear equation system is just built in a fixed way (SVD, Chebeyshev, etc.). The user has usually no choice to optimize conditioning depending on the "soundness" of data (amount of noise). However, there is a compromise in "solution stability" versus "resolution". I wanted to give the user the opportunity to try conditioning for the highest possible resolution (i.e., no conditioning or lowest possible conditioning. There is no mathematical straight forward way to choose. The user can tweak rcond to get a "reasonable" solution with acceptable resolution. Increasing rcond smoothes the result, lowering it beyond a minimum renders the solution unstable (i.e., you get insane solutions with values 10^17, still calculating back the "solutions" solve the equation system). The intention now is to let the user tweak on recond until he gets the best compromise.